### PR TITLE
[Schedule] Fix linear bias after row sharding weight (#10)

### LINKED ISSUE #10: [Bug] Multiple reductions of linear bias after row sharding the weight
Consider the following case, we have partial inputs and shard the weight $A$ of the linear layer by row.

$$
XA+b
=\begin{bmatrix}X_1 & X_2\end{bmatrix}
\begin{bmatrix}A_1 \\\\ A_2\end{bmatrix}
+b
= X_1A_1+X_2A_2+b
$$

In the current implementation (as shown below), we directly conduct `all_reduce` for the output of the linear layer. However, since the bias $b$ is implicitly included in the linear layer, this leads to redundant additions for the bias and generates incorrect results.

```python
class Model(nn.Module):
  def __init__(self):
    self.linear = nn.linear(...)
  def forward(self, x):
    x = self.linear(x)
    dist.all_reduce(x, op=dist.ReduceOp.SUM)
    return x
```

As a result, we need to preprocess the bias when the corresponding linear layer is row sharded. There are two possible solutions:
1. Only one device holds the bias data and other devices set bias to None, so the bias will only be added once.
2. Average the bias across different devices. In that case, the summation of the new bias will be the same as the original one.

### PR DESCRIPTION
**Pull request by chhzh123**
## Description ##
This PR fixes #10. It leverages Megatron-LM's method to delay the bias add after `all_reduce`, so that the bias will only be added once, and the gradients are proved to be the same as the original ones.

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

**Comment by chhzh123** [comment]
But the sequence parallelism case have not passed yet. I'm not sure why `linear2.bias.grad` is 1/TP of the original one, though I only add the bias once.

File: slapo/op/linear.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        factory_kwargs = {"device": device, "dtype": dtype}
+        super().__init__(
+            in_features, out_features, bias=False, device=device, dtype=dtype
+        )
+        self.sync_fn = sync_fn
+        self.bias = Parameter(torch.empty(out_features, **factory_kwargs))
+        self.reset_parameters()
+
+    def forward(self, x: Tensor) -> Tensor:  # pylint: disable=arguments-renamed
+        x = F.linear(x, self.weight, None)
```
Comment (comaniac) [inline slapo/op/linear.py:35]: Can you just call the forward function of the base class (i.e., linear)? I suppose it should have the same functionality, but might be more clean.

File: slapo/op/linear.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        # pylint: disable=comparison-with-callable
+        if self.sync_fn.func == dist.all_reduce:
+            self.sync_fn(x)
+        else:
+            x = self.sync_fn(x)
```
Comment (comaniac) [inline slapo/op/linear.py:40]: Add a comment saying what this sync function could be. It would be good to also add a complete docstring including arguments for this class.

File: slapo/schedule.py [hunk 3]
```diff
@@ -428,6 +430,26 @@ def validate_sync_op(mode, sync_op_or_fn, axis=None):
...
+                        ):
+                            new_mod = LinearWithSeparateBias(
+                                self.mod.in_features,
+                                self.mod.out_features,
+                                sync_fn,
+                                self.mod.weight.device,
+                                self.mod.weight.dtype,
+                            )
+                            self.replace(new_mod)
+                        return
```
Comment (comaniac) [inline slapo/schedule.py:452]: Some concerns:
1. This breaks tie weights.
2. The execution flow of .shard becomes a bit difficult to flow, because here is an early return. I'm not sure if there's a better way but I'll think about it.

**Comment by szhengac** [comment]
> But the sequence parallelism case has not passed yet. I'm not sure why `linear2.bias.grad` is 1/TP of the original one, though I only add the bias once.

As I mentioned offline, this is because each bias only pass through 1/TP tokens due to sequence dimension partitioning. We need to discuss it more before we can merge this pr.

File: slapo/op/linear.py (hunk 1 above)
Comment (chhzh123) [reply slapo/op/linear.py:40]: I removed the previous bias creation and `reset_parameters`, but retain the `F.linear` function. Should have the same functionality.

File: slapo/op/linear.py (hunk 2 above)
Comment (chhzh123) [reply slapo/op/linear.py]: Fixed

File: slapo/schedule.py (hunk 3 above)
Comment (chhzh123) [reply slapo/schedule.py]: Added `tie_weights` support

File: slapo/schedule.py [hunk 4]
```diff
@@ -407,6 +409,44 @@ def validate_sync_op(mode, sync_op_or_fn, axis=None):
...
+                            "but they have different sharded shapes: "
+                            f"{new_tensor.shape} vs "
+                            f"{sch.metadata.tie_weights[param].shape}"
+                        )
+                    new_param = sch.metadata.tie_weights[param]
+                else:
+                    new_param = new_mod.weight
+                    self.metadata.tie_weights[param] = new_param
+            else:
+                new_param = new_mod.weight
```
Comment (comaniac) [inline slapo/schedule.py:447]: Could we have a utility function for this logic so that we could just call it when needed?

Comment (chhzh123) [reply slapo/schedule.py]: Added

**Comment by chhzh123** [comment]
> But the sequence parallelism case has not passed yet. I'm not sure why `linear2.bias.grad` is 1/TP of the original one, though I only add the bias once.

Extra [parameter hook](https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html) is added to make sure the bias grad is correct.

File: slapo/schedule.py [hunk 5]
```diff
@@ -219,6 +224,24 @@ def __contains__(self, full_path):
                 return False
         return True
 
+    def get_params_if_tied(self, old_param, new_param):
```
Comment (comaniac) [inline slapo/schedule.py:227]: The function name seems not clear to me. Maybe "get_or_set_tied_param" or something like that.

File: slapo/schedule.py [hunk 6]
```diff
@@ -1093,6 +1142,18 @@ def init_target_engine(sch, target, **kwargs):
     )
 
 
+def reduce_model_grads(sch: Schedule):
```
Comment (comaniac) [inline slapo/schedule.py:1145]: I like this idea, but not sure if there's any issue with this implementation. Maybe we could revisit it later.

**Review by comaniac — APPROVED** [review]
LGTM. Just nits. I'll leave this PR to @szhengac for merging.

File: slapo/schedule.py [hunk 7]
```diff
@@ -1154,6 +1160,18 @@ def init_target_engine(sch, target, **kwargs):
...
+    def hook(grad):
+        dist.all_reduce(grad.contiguous(), dist.ReduceOp.SUM, group=sch.group)
```
Comment (szhengac) [inline slapo/schedule.py:1147]: This is not an efficient way to allreduce bias, since there are hundreds of bias in a large model. and we will end up with large communication latency, greatly slowing down the training. The best way for bias will be merging them in the end to perform a single allreduce call. for example, take a look at where layernorm gradient is reduced in megatron https://github.com/NVIDIA/Megatron-LM/blob/master/megatron/optimizer/optimizer.py#L285

Comment (chhzh123) [reply slapo/schedule.py:1147]: Right, I've seen the Megatron implementation, but the problem is that it is implemented in optimizer, so it can all_reduce the parameters in one place. For our case, we only return a model for users, and the computation and communication should be included in the model, so that's why I use param hooks instead.

Comment (szhengac) [reply slapo/schedule.py:1147]: you can invoke gradient sync after the whole stage backward. nonetheless, we cannot use the current way of synchronizing gradients otherwise we will end up with extremely slow training example.

File: slapo/schedule.py (hunk 5 above)
Comment (chhzh123) [reply slapo/schedule.py]: Fixed

**Comment by chhzh123** [comment]
Moved gradient reduction to the end of each backward pass per @szhengac 's suggestion.

**Comment by comaniac** [comment]
Closed as #74 already covers this feature.
